addons/website_sale/tests/test_dynamic_snippet_category.py

```diff
@@ -43,17 +50,63 @@ def test_snippet_categories_returns_only_published_and_with_children(self):
 
     def test_set_category_image(self):
         """Test setting a cover image via JSON-RPC route"""
-        attachment = self.env['ir.attachment'].create({
-            'name': "test.png",
-            'datas': 'iVBORw0KGgoAAAANSUhEUgAAAAYAAAAGCAYAAADgzO9IAAAAJElEQVQI'
-                     'mWP4/b/qPzbM8Pt/1X8GBgaEAJTNgFcHXqOQMV4dAMmObXXo1/BqAAAA'
-                     'AElFTkSuQmCC',
-            'public': True,
-        })
         with MockRequest(self.website.env, website=self.website):
-            self.website_sale.set_category_image(self.category1.id, attachment.id)
+            self.website_sale.set_category_image(self.category1.id, self.test_attachment.id)
             self.assertEqual(
                 self.category1.cover_image,
-                attachment.datas,
+                self.test_attachment.datas,
                 "Cover image should match the uploaded attachment",
             )
+
+    def test_set_category_image_portal_user(self):
+        res = self.url_open(
+            '/snippets/category/set_image',
+            data={'category_id': self.category1.id, 'attachment_id': self.test_attachment.id},
+            method='POST',
+        )
+        self.assertEqual(
+            res.status_code,
+            415,
+            "Portal users should not be able to set category images",
+        )
+
+    def test_prepare_category_list_data(self):
+        """Test the data preparation for the category list"""
+        with MockRequest(self.website.env, website=self.website):
+            response = self.env['website.snippet.filter']._prepare_category_list_data()
+            base_url = self.website.get_base_url()
+            default_img_path = self.env['product.template']._get_product_placeholder_filename()
+            default_img_url = f'{base_url}/{default_img_path}'
+            expected_response = [{
+                'id': self.category1.id,
+                'name': "Published Category",
+                'unpublished': not self.category1.has_published_products,
+                'cover_image': default_img_url,
+            }, {
+                'id': self.category2.id,
+                'name': "Published Category 2",
+                'unpublished': not self.category2.has_published_products,
+                'cover_image': default_img_url,
+            }]
```

Comment:
we could make function for the

```
def get_expected_response_dict(category):
    return {
        'id': category.id,
        'name': category.name,
        'unpublished': not category.has_published_products,
        'cover_image': (
                f'{base_url}{request.website.image_url(cat, "cover_image")}'
                if category.cover_image else default_img_url
            ),
    }
```

and in method call

```
    categories = CategorySudo.search(domain & Domain('parent_id', '=', False))
    expected_response = [get_expected_response_dict(cat) for cat in categories]
    self.assertEqual(expected_response, response)
```
Why not assertEqual? 🤔 

same method can be used for below